File: test-suite-generator/src/test/java/io/micronaut/jsonschema/generator/animals/AnimalTest.java
/*
 * Copyright 2017-2024 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.jsonschema.generator.animals;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.json.JsonMapper;
import org.junit.jupiter.api.Test;

import static org.junit.Assert.assertEquals;

public class AnimalTest {

    JsonMapper jsonMapper = new JsonMapper();

    @Test
    public void mapAnimal() throws JsonProcessingException {
        var animal = jsonMapper.readValue("""
            {
              "id": "0x",
              "birthdate": "2000-01-01",
              "name": "Micronaut",
              "resourceType": "Animal"
            }
            """, Animal.class);
        assertEquals(Animal.class, animal.getClass());
        assertEquals("0x", animal.getId());
        assertEquals("2000-01-01", animal.getBirthdate());
        assertEquals("Micronaut", animal.getName());
    }

    @Test
    public void mapCat() throws JsonProcessingException {
        var animal = jsonMapper.readValue("""
            {
              "id": "0x",
              "birthdate": "2000-01-01",
              "name": "Micronaut",
              "resourceType": "Cat",
              "hasMate": true
            }
            """, Animal.class);
        assertEquals(Cat.class, animal.getClass());
        Cat cat = (Cat) animal;

        assertEquals("0x", cat.getId());
        assertEquals("2000-01-01", cat.getBirthdate());
        assertEquals("Micronaut", cat.getName());
        assertEquals(true, cat.getHasMate());
        assertEquals("Cat", cat.resourceType);
    }

    @Test
    public void mapDog() throws JsonProcessingException {
        var animal = jsonMapper.readValue("""
            {
              "id": "0x",
              "birthdate": "2000-01-01",
              "name": "Micronaut",
              "resourceType": "Dog",
              "hasMate": true,
              "nickname": "Goodie",
              "ownerName": "Owner",
              "enemies": [
              {
                  "id": "1x",
                  "birthdate": "2000-01-01",
                  "name": "Pegasus",
                  "resourceType": "Cat"
              }, {
                  "id": "2x",
                  "birthdate": "2000-01-01",
                  "name": "Micro-Pego",
                  "resourceType": "Cat"
              }]
            }
            """, Animal.class);
        assertEquals(Dog.class, animal.getClass());
        Dog dog = (Dog) animal;

        assertEquals("0x", dog.getId());
        assertEquals("2000-01-01", dog.getBirthdate());
        assertEquals("Micronaut", dog.getName());
        assertEquals(true, dog.getHasMate());
        assertEquals("Dog", dog.resourceType);
        assertEquals("Goodie", dog.getNickname());
        assertEquals("Owner", dog.getUnknownFields().get("ownerName"));
        assertEquals(2, dog.getEnemies().size());
    }

    @Test
    public void mapFish() throws JsonProcessingException {
        var animal = jsonMapper.readValue("""
            {
              "id": "0x",
              "birthdate": "2000-01-01",
              "name": "Micronaut",
              "resourceType": "Fish",
              "ownerName": "Owner",
              "friends": [
              {
                  "id": "1x",
                  "birthdate": "2000-01-01",
                  "name": "Pegasus",
                  "resourceType": "Fish",
                  "ownerName": "Owner"
              }, {
                  "id": "2x",
                  "birthdate": "2000-01-01",
                  "name": "Micro-Pego",
                  "resourceType": "Fish",
                  "ownerName": "Owner"
              }]
            }
            """, Animal.class);
        assertEquals(Fish.class, animal.getClass());
        Fish fish = (Fish) animal;

        assertEquals("0x", fish.getId());
        assertEquals("2000-01-01", fish.getBirthdate());
        assertEquals("Micronaut", fish.getName());
        assertEquals("Fish", fish.resourceType);
        assertEquals("Owner", fish.otherFields().get("ownerName"));
        assertEquals(2, fish.getFriends().size());
    }

    @Test
    public void mapHuman() throws JsonProcessingException {
        var animal = jsonMapper.readValue("""
            {
              "id": "N-XN",
              "birthdate": "2000-01-01",
              "name": "ABCDEFGHIJKLMNOPQRSTUVW",
              "resourceType": "Human",
              "firstName": "ABCDEFGHIJKLMNO",
              "lastName": "ABCD",
              "sport": "ABCDE"
            }
            """, Animal.class);
        assertEquals(Human.class, animal.getClass());
        Human human = (Human) animal;
        assertEquals("N-XN", human.getId());
        assertEquals("2000-01-01", human.getBirthdate());
        assertEquals("Human", human.resourceType);
        assertEquals("ABCDEFGHIJKLMNO", human.getFirstName());
    }
    // TODO: enum test
}

File: test-suite-generator-kotlin/src/test/kotlin/io/micronaut/jsonschema/generator/animals/AnimalTest.kt
package io.micronaut.jsonschema.generator.animals

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import tools.jackson.databind.json.JsonMapper

class AnimalTest {

    val jsonMapper = JsonMapper()

    @Test
    fun mapAnimal() {
        val animal = jsonMapper.readValue(
            """
            {
              "id": "0x",
              "birthdate": "2000-01-01",
              "name": "Micronaut",
              "resourceType": "Animal"
            }
            """, Animal::class.java
        )
        assertEquals(Animal::class.java, animal.javaClass)
        assertEquals("0x", animal.id)
        assertEquals("2000-01-01", animal.birthdate)
        assertEquals("Micronaut", animal.name)
    }

    // tag::mapp[]
    @Test
    fun mapCat() {
        val animal = jsonMapper.readValue(
            """
            {
              "id": "0x",
              "birthdate": "2000-01-01",
              "name": "Micronaut",
              "resourceType": "Cat",
              "hasMate": true
            }
            """, Animal::class.java
        )
        assertEquals(Cat::class.java, animal.javaClass)
        val cat = animal as Cat

        assertEquals("0x", cat.id)
        assertEquals("2000-01-01", cat.birthdate)
        assertEquals("Micronaut", cat.name)
        assertEquals(true, cat.hasMate)
        assertEquals("Cat", cat.resourceType)
    }
    // end::mapp[]

    @Test
    fun mapDog() {
        val animal = jsonMapper.readValue(
            """
            {
              "id": "0x",
              "birthdate": "2000-01-01",
              "name": "Micronaut",
              "resourceType": "Dog",
              "hasMate": true,
              "nickname": "Goodie",
              "ownerName": "Owner",
              "enemies": [
              {
                  "id": "1x",
                  "birthdate": "2000-01-01",
                  "name": "Pegasus",
                  "resourceType": "Cat"
              }, {
                  "id": "2x",
                  "birthdate": "2000-01-01",
                  "name": "Micro-Pego",
                  "resourceType": "Cat"
              }]
            }
            """, Animal::class.java
        )
        assertEquals(Dog::class.java, animal.javaClass)
        val dog = animal as Dog

        assertEquals("0x", dog.id)
        assertEquals("2000-01-01", dog.birthdate)
        assertEquals("Micronaut", dog.name)
        assertEquals(true, dog.hasMate)
        assertEquals("Dog", dog.resourceType)
        assertEquals("Goodie", dog.nickname)
        assertEquals("Owner", dog.unknownFields.getOrDefault("ownerName", ""))
        assertEquals(2, dog.enemies.size)
    }

    @Test
    fun mapFish() {
        val animal = jsonMapper.readValue(
            """
            {
              "id": "0x",
              "birthdate": "2000-01-01",
              "name": "Micronaut",
              "resourceType": "Fish",
              "ownerName": "Owner",
              "friends": [
              {
                  "id": "1x",
                  "birthdate": "2000-01-01",
                  "name": "Pegasus",
                  "resourceType": "Fish",
                  "ownerName": "Owner"
              }, {
                  "id": "2x",
                  "birthdate": "2000-01-01",
                  "name": "Micro-Pego",
                  "resourceType": "Fish",
                  "ownerName": "Owner"
              }]
            }
            """, Animal::class.java
        )
        assertEquals(Fish::class.java, animal.javaClass)
        val fish = animal as Fish

        assertEquals("0x", fish.id)
        assertEquals("2000-01-01", fish.birthdate)
        assertEquals("Micronaut", fish.name)
        assertEquals("Fish", fish.resourceType)
        assertEquals("Owner", fish.unknownFields["ownerName"])
        assertEquals(2, fish.friends.size)
    }

    @Test
    fun mapHuman() {
        val animal = jsonMapper.readValue(
            """
            {
              "id": "N-XN",
              "birthdate": "2000-01-01",
              "name": "ABCDEFGHIJKLMNOPQRSTUVW",
              "resourceType": "Human",
              "firstName": "ABCDEFGHIJKLMNO",
              "lastName": "ABCD",
              "sport": "ABCDE"
            }
            """, Animal::class.java
        )
        assertEquals(Human::class.java, animal.javaClass)
        val human = animal as Human
        assertEquals("N-XN", human.id)
        assertEquals("2000-01-01", human.birthdate)
        assertEquals("Human", human.resourceType)
        assertEquals("ABCDEFGHIJKLMNO", human.firstName)
    }
}
